src/flint/test/test_all.py

```diff
@@ -2631,6 +2633,398 @@ def setbad(obj, i, val):
+        assert raises(lambda: P(ctx=ctx) <= None, TypeError)
+        assert raises(lambda: P(ctx=ctx) > None, TypeError)
+        assert raises(lambda: P(ctx=ctx) >= None, TypeError)
+        assert raises(lambda: None < P(ctx=ctx), TypeError)
+        assert raises(lambda: None <= P(ctx=ctx), TypeError)
+        assert raises(lambda: None > P(ctx=ctx), TypeError)
+        assert raises(lambda: None >= P(ctx=ctx), TypeError)
+
+        p = quick_poly()
+        assert p.coefficient(2) == S(2)
+        assert raises(lambda: p.coefficient(-1), IndexError)
+        assert raises(lambda: p.coefficient(10), IndexError)
+
+        assert p[0] == mpoly({(2, 2): 4})
+        assert p[3] == mpoly({(0, 0): 1})
+        assert raises(lambda: p[-1], IndexError)
+        assert raises(lambda: p[4], IndexError)
+
+        assert p[(2, 2)] == 4
+        assert p[(0, 0)] == 1
+        assert raises(lambda: p[(1,)], ValueError)
+        assert raises(lambda: p[(1, "bad")], TypeError)
+        assert raises(lambda: p["bad"], TypeError)
+
+        p = quick_poly()
+        p[1] = S(10)
+        assert p == mpoly({(0, 0): 1, (0, 1): 2, (1, 0): 10, (2, 2): 4})
+
+        p = quick_poly()
+        p[(1, 0)] = S(10)
+        assert p == mpoly({(0, 0): 1, (0, 1): 2, (1, 0): 10, (2, 2): 4})
+
+        assert raises(lambda: p.__setitem__(-1, 1), IndexError)
+        assert raises(lambda: p.__setitem__(4, 1), IndexError)
+
+        assert raises(lambda: p.__setitem__((1,), 1), ValueError)
+        assert raises(lambda: p.__setitem__((1, "bad"), 1), TypeError)
+        assert raises(lambda: p.__setitem__("bad", 1), TypeError)
+
+        assert raises(lambda: p.__setitem__(2, None), TypeError)
+        assert raises(lambda: p.__setitem__(-1, 1), IndexError)
+
+        assert P(ctx=ctx).repr() == f"{ctx.__class__.__name__}(2, '<Ordering.lex: 0>', ('x0', 'x1')).from_dict({{}})"
+        assert P(1, ctx=ctx).repr() == f"{ctx.__class__.__name__}(2, '<Ordering.lex: 0>', ('x0', 'x1')).from_dict({{(0, 0): 1}})"
+        assert str(quick_poly()) == repr(quick_poly()) == '4*x0^2*x1^2 + 3*x0 + 2*x1 + 1'
+
+        assert p.exponent_tuple(0) == (2, 2)
+        assert p.exponent_tuple(3) == (0, 0)
+        assert raises(lambda: p.exponent_tuple(-1), IndexError)
+        assert raises(lambda: p.exponent_tuple(4), IndexError)
+
+        assert p.total_degree() == 4
+        assert P(ctx=ctx).total_degree() == -1
+        assert P(1, ctx=ctx).total_degree() == 0
+
+        p = quick_poly()
+        assert p(0, 0) == p(0, S(0)) == p(S(0), S(0)) == p(x1=S(0), x0=S(0)) == S(1) == 1
+        assert p(1, 1) == p(x1=S(1), x0=S(1)) == S(10) == 10
+        assert p(x0=p(x1=0)) == mpoly({
+            (2, 2): 36,
+            (1, 2): 24,
+            (1, 0): 9,
+            (0, 2): 4,
+            (0, 1): 2,
+            (0, 0): 4
+        })
+        assert p() == p
+        assert p(x0=ctx.from_dict({(1, 0): 1}), x1=ctx.from_dict({(0, 1): 1})) == p
```

Comment:
I think it would be better if `__call__` is restricted to being called with the right number of arguments and only takes positional arguments.

For partial evaluation it would be better to have a different method like: `p.subs({'x0': 1})`.

Reply:
I'm also unsure about mixing evaluation and composition here. Maybe it would be better if there was a separate `compose` method. It could take positional arguments all being polynomials in the same ring but not necessarily the ring of the calling polynomial e.g.:
```python
In [5]: from flint import *

In [6]: ctx1 = fmpq_mpoly_ctx.get_context(2, Ordering.lex, 'x')

In [7]: ctx2 = fmpq_mpoly_ctx.get_context(2, Ordering.lex, 'y')

In [8]: p = fmpq_mpoly('x0*2 + x1*3', ctx1)

In [9]: p2 = fmpq_mpoly('y0*2 + y1*3', ctx2)

In [10]: p.compose(p2, p2)
```

Reply:
From a typing perspective I think it is better if we are clear that `p(...)` returns `fmpq` and `p.subs(...)` or `p.compose(...)` return `fmpq_mpoly`.

Reply:
I've split both call methods up into a
- `__call__`: accepts only positional constant arguments, returns a constant
- `subs`: accepts a dict of variable names to constants, returns a mpoly
- `compose`: accepts only positional mpoly arguments, returns a mpoly. All arguments must share the same context.